text/0018-message-ordering-guarantee.md

```diff
@@ -0,0 +1,175 @@
+
+3. Perform part 1 of the refactor without part 2 - that is, consolidate the message sending, but leave the RPCs defaulted to the Update stage. This has less chance of causing user friction, at the cost of accepting the potential that the default update stage of `Update` for RPCs could create edge cases that are not intuitive to the user where they would be required to set the `UpdateStage` value in their send parameters. As using send parameters at all is somewhat more advanced than the most basic usage of RPCs, the assertion is made that the fewer surprises we can allow into the basic usage, the more intuitive that usage will be; the `UpdateStage` parameter will remain for advanced users, but an intelligent default value should diminish the need for users to interact with that parameter.
+
+4. We discussed integrating this change into the current snapshotting system proposal. There are two reasons this isn't currently achievable:
+   
+   - The snapshotting system, as currently proposed, only sends delta value updates and doesn't have a mechanism for carrying commands, and
+   
+   - The snapshotting system, as currently proposed, sends all of its messages unreliably, which introduces the possibility of edge cases where dropped packets unexpectedly change the order of commands.
+   
+   However, as mentioned above, when the snapshotting system is better realized to convey these sorts of messages, the consolidation of the message sending logic into a single mechanism should still be of benefit to the integration with snapshotting, as it will reduce the number of different types of messaging logic the snapshotting system needs to integrate down to just one.
+
+# Prior art
+
+This modification is fairly MLAPI- and Unity-specific. I'm unaware of any existing implementations that attempt to solve this particular problem.
+
+# Unresolved questions
+
+- Currently, there are two different message type values being passed through the system: the `RpcQueueContainer.QueueItemType` value and the value from `NetworkConstants.cs`. It appears the primary reason for this separation was to be able to identify which messages contained the queue item message structure, and which didn't, but is there still value to this separation? Or is there more value now in keeping only one and removing the other entirely?
+- Should internal commands that are triggered by users (such as `Spawn()`, `Despawn()`, and `ChangeOwnership()`) gain optional parameters to allow the user to control which stage they execute on?
+
+# Future possibilities
+
+As mentioned above in the rationale section, there is a known intention to integrate commands into the snapshot system. While this RFC doesn't attempt to address that yet, the consolidation of message sending mechanisms can be seen as a first step toward realizing that.
```

Comment:
I think it's all good — I might try hard to nit-pick but this RFC, even though not being the ultimate solution to our queuing (or in wider terms, network input/output loop), it still fixes some issues and adds immediate value in the short term.

I might only suggest mentioning that all this work could be refactored or even potentially rewritten when we clearly define a network frame and network input/output loop. having said that, such changes would require other redesign/rework on the transport API abstraction and use of Unity.Collections more — just wanted to plug here instead of just saying LGTM 🚀 :)